File: core/models/report.py
"""Cross-platform reporting schema."""
from dataclasses import dataclass, field
from datetime import date
from typing import Optional
from .metrics import NormalizedMetrics

@dataclass
class CrossPlatformReport:
    """Aggregated report across multiple platforms."""
    date_start: date
    date_end: date
    platforms: list[str] = field(default_factory=list)
    metrics_by_platform: dict[str, list[NormalizedMetrics]] = field(default_factory=dict)

    @property
    def total_spend(self) -> float:
        return sum(m.cost for metrics in self.metrics_by_platform.values() for m in metrics)

    @property
    def total_conversions(self) -> float:
        return sum(m.conversions for metrics in self.metrics_by_platform.values() for m in metrics)

    @property
    def blended_cpa(self) -> float:
        tc = self.total_conversions
        return self.total_spend / tc if tc > 0 else 0.0

File: dashboard/src/types/report.ts
import type { NormalizedMetrics } from './metrics';

export interface CrossPlatformReport {
  dateStart: string;
  dateEnd: string;
  platforms: string[];
  metricsByPlatform: Record<string, NormalizedMetrics[]>;
}

export function totalSpend(report: CrossPlatformReport): number {
  return Object.values(report.metricsByPlatform)
    .flat()
    .reduce((sum, m) => sum + m.cost, 0);
}

export function totalConversions(report: CrossPlatformReport): number {
  return Object.values(report.metricsByPlatform)
    .flat()
    .reduce((sum, m) => sum + m.conversions, 0);
}

export function blendedCpa(report: CrossPlatformReport): number {
  const tc = totalConversions(report);
  return tc > 0 ? totalSpend(report) / tc : 0;
}
